### PR TITLE
Migrate to ReCaptcha's v2 API

### PR DESCRIPTION
**Pull request by akrabat**
Update @lansoweb's PR (#8) to ensure MailHide still works with its v1 API as it's not been updated to v2 by Google.

Also ensure Travis tests on PHP 5.6, 7, and 7.1. For v2, Google provide test keys, so we can use them and let Travis test better.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -10,21 +10,20 @@ cache:
 matrix:
   fast_finish: true
   include:
-    - php: 5.5
```
Comment (weierophinney) [inline .travis.yml]: :tada:

File: README.md [hunk 2]
```diff
@@ -7,7 +7,7 @@ You can install using:
 ```
 curl -s https://getcomposer.org/installer | php
 php composer.phar install
-composer require zendframework/zendservice-recaptcha
+php composer.phar require zendframework/zendservice-recaptcha
 ```
```
Comment (weierophinney) [inline README.md]: Alternately, we point them to the Composer download page, and demonstrate using `composer require` (not usage of the `phar` file). Either way works, though we've been trending towards just saying `composer require`.

File: src/MailHide.php [hunk 3]
```diff
@@ -186,15 +186,45 @@ public function getDefaultOptions()
...
+
+    /**
+     * Get the public key property
+     *
+     * @return string
+     */
+    public function getPublicKey()
+    {
+        return parent::getSiteKey();
+    }
```
Comment (weierophinney) [inline src/MailHide.php]: Should these proxy methods maybe be marked deprecated now?

File: src/MailHide.php [hunk 4]
```diff
@@ -273,11 +303,11 @@ public function getHtml($email = null)
             throw new MailHideException('Missing email address');
         }
 
-        if ($this->publicKey === null) {
+        if ($this->getPublicKey() === null) {
```
Comment (weierophinney) [inline src/MailHide.php]: Why `getPublicKey()` and not `getSiteKey()`?

File: src/MailHide.php [hunk 5]
```diff
@@ -273,11 +303,11 @@ public function getHtml($email = null)
...
             throw new MailHideException('Missing public key');
         }
 
-        if ($this->privateKey === null) {
+        if ($this->getPrivateKey() === null) {
```
Comment (weierophinney) [inline src/MailHide.php]: Why `getPrivateKey()` and not `getSecretKey()`?

File: src/ReCaptcha.php [hunk 6]
```diff
@@ -135,7 +123,7 @@ public function __construct(
             $this->setOptions($options);
         }
 
-        $this->setHttpClient($httpClient ?: new HttpClient);
+        $this->setHttpClient($httpClient ?  : new HttpClient());
```
Comment (weierophinney) [inline src/ReCaptcha.php]: No space between `?` and `:` here, please; this should be the shortened ternary form for "assign default value only if test value is not truthy".

File: src/ReCaptcha.php [hunk 7]
```diff
@@ -174,7 +162,7 @@ public function __toString()
      * Set the ip property
      *
      * @param string $ip
-     * @return ReCaptcha
+     * @return \ZendService\ReCaptcha\ReCaptcha
```
Comment (weierophinney) [inline src/ReCaptcha.php]: `@return self`

File: src/ReCaptcha.php [hunk 8]
```diff
@@ -198,7 +186,7 @@ public function getIp()
      *
      * @param string $key
      * @param string $value
-     * @return ReCaptcha
+     * @return \ZendService\ReCaptcha\ReCaptcha
```
Comment (weierophinney) [inline src/ReCaptcha.php]: here as well, and throughout.

File: src/ReCaptcha.php [hunk 9]
```diff
@@ -210,21 +198,21 @@ public function setParam($key, $value)
     /**
      * Set parameters
      *
-     * @param  array|Traversable $params
-     * @return ReCaptcha
-     * @throws Exception
+     * @param array|Traversable $params
+     * @return \ZendService\ReCaptcha\ReCaptcha
+     * @throws \ZendService\ReCaptcha\Exception
```
Comment (weierophinney) [inline src/ReCaptcha.php]: Can be just `Exception`, as it's in the same namespace.

File: src/ReCaptcha.php [hunk 10]
```diff
@@ -368,85 +354,93 @@ public function setPrivateKey($privateKey)
...
-            $host = self::API_SECURE_SERVER;
-        }
-
-        $htmlBreak = '<br>';
-        $htmlInputClosing = '>';
-
-        if ((bool) $this->params['xhtml'] === true) {
-            $htmlBreak = '<br />';
-            $htmlInputClosing = '/>';
+        // Should We use a onload callback?
```
Comment (weierophinney) [inline src/ReCaptcha.php]: `s/ a / an /`

File: src/ReCaptcha.php [hunk 11]
```diff
@@ -368,85 +354,93 @@ public function setPrivateKey($privateKey)
...
+        if (!empty($this->options['onload'])) {
```
Comment (weierophinney) [inline src/ReCaptcha.php]: Space following `!` please (for internal consistency; you've done that everywhere else I've seen! :smile:)

File: src/ReCaptcha.php [hunk 12]
```diff
@@ -368,85 +354,93 @@ public function setPrivateKey($privateKey)
...
+                $this->options['onload']
+            );
         }
 
-        $errorPart = '';
+        $langOption = '';
 
-        if (!empty($this->params['error'])) {
-            $errorPart = '&error=' . urlencode($this->params['error']);
+        if (!empty($this->options['hl'])) {
```
Comment (weierophinney) [inline src/ReCaptcha.php]: Space following `!` operator, please.

File: composer.json [hunk 13]
```diff
@@ -18,22 +18,16 @@
...
-        "zendframework/zend-http": "~2.0",
-        "zendframework/zend-json": "~2.0"
+        "php": "^5.6 || ^7.0",
+        "zendframework/zend-http": "^2.5.4",
+        "zendframework/zend-json": "^2.6.1 || ^3.0"
     },
     "require-dev": {
-        "phpunit/phpunit": "~4.0",
+        "phpunit/phpunit": "^5.7 || ^6.0",
         "squizlabs/php_codesniffer": "^2.3",
```
Comment (weierophinney) [inline composer.json]: Change this dependency to `"zendframework/zend-coding-standard": "~1.0.0"`, please, and then re-run phpcs and/or phpcbf; they'll pick up the various CS bits I've noticed.

File: src/Response.php [hunk 14]
```diff
@@ -23,30 +23,30 @@ class Response
...
+    protected $errorCodes = [];
 
     /**
      * Class constructor used to construct a response
      *
      * @param string $status
-     * @param string $errorCode
-     * @param HTTPResponse $httpResponse If this is set the content will override $status and $errorCode
+     * @param array $errorCodes
+     * @param \Zend\Http\Response $httpResponse If this is set the content will override $status and $errorCode
```
Comment (weierophinney) [inline src/Response.php]: Use the same typehint as found in the signature; IDEs and phpdoc will expand them correctly.

File: src/Response.php [hunk 15]
```diff
@@ -57,16 +57,12 @@ public function __construct($status = null, $errorCode = null, HTTPResponse $htt
     /**
      * Set the status
      *
-     * @param string $status
-     * @return Response
+     * @param boolean $status
+     * @return \ZendService\ReCaptcha\Response
```
Comment (weierophinney) [inline src/Response.php]: `@return self`

File: src/Response.php [hunk 16]
```diff
@@ -57,16 +57,12 @@ public function __construct($status = null, $errorCode = null, HTTPResponse $htt
     /**
      * Set the status
      *
-     * @param string $status
-     * @return Response
+     * @param boolean $status
```
Comment (weierophinney) [inline src/Response.php]: `bool` instead of `boolean` (internal consistency)

File: test/ReCaptchaTest.php [hunk 17]
```diff
@@ -20,18 +21,25 @@ class ReCaptchaTest extends TestCase
...
-            || $this->publicKey == 'public key'
-            || empty($this->privateKey)
-            || $this->privateKey == 'private key'
-        ) {
-            $this->markTestSkipped('ZendService\ReCaptcha\ReCaptcha tests skipped due to missing keys');
+        $this->siteKey = getenv('TESTS_ZEND_SERVICE_RECAPTCHA_SITE_KEY');
+        $this->secretKey = getenv('TESTS_ZEND_SERVICE_RECAPTCHA_SECRET_KEY');
+
+        if (empty($this->siteKey)) {
+            $this->siteKey = '6LeIxAcTAAAAAJcZVRqyHh71UMIEGNQ_MXjiZKhI';
```
Comment (weierophinney) [inline test/ReCaptchaTest.php]: Maybe these should be in the `phpunit.xml.dist` as the default values, with a comment above indicating users should change them if they want to use their own keys? That would simplify the test code a bit.

File: src/MailHide.php (hunk 4 above)
Comment (akrabat) [inline src/MailHide.php]: MailHide is still on v1's API which uses the nomenclature "public_key" and "private_key".

File: src/MailHide.php (hunk 5 above)
Comment (akrabat) [inline src/MailHide.php]: MailHide is still on v1's API which uses the nomenclature "public_key" and "private_key".

File: src/MailHide.php (hunk 3 above)
Comment (akrabat) [inline src/MailHide.php]: They aren't actually deprecated for MailHide as that's the correct wording for its API.

**Comment by akrabat** [comment]
@weierophinney Review items addressed

File: src/MailHide.php (hunk 3 above)
Comment (weierophinney) [inline src/MailHide.php]: Okay, sounds good.

File: src/MailHide.php (hunk 4 above)
Comment (weierophinney) [inline src/MailHide.php]: Makes sense. Maybe add some verbiage indicating that, then?

**Comment by xorock** [comment]
Hi guys. Some time ago I wrote full service for recaptcha v2 https://github.com/xorock/zend-service-recaptcha-v2. The only thing that is missing here are tests.

**Pull request merged**